mozilla-services/socorro
# bug 1703252: upgrade to django 3.2

This upgrades from Django 2.2 to Django 3.2. It fixes deprecations and a variety of other things.

* 8a51917da bug 1703252: swap out python-memcached for pymemcache
* 8b56a49e2 Fix importing from collections
* 17690a90e bug 1703252: clean up settings
* 6363facda bug 1703252: fix middleware tests
* 9d47cf71e bug 1703252: upgrade django-waffle to 2.3.0
* 297f9965f bug 1703252: no longer need to define default_app_config
* b8e235b91 bug 1703252: convert smart_text to smart_str
* 2f1c2ef69 bug 1703252: fix django-pipeline and autofield
* a19925505 bug 1703252: convert django.utils.http.urlquote to urllib.parse.quote
* 93e22562f bug 1703252: switch unique_together and ordering to lists
* d82e6393e bug 1703252: remove deprecated use of is_ajax
* 90bdac792 bug 1703252: remove NullBooleanField
* 97ee13f2a bug 1703252: convert url to re_path

To test:

1. `make build`
2. `make lint` and `make lintfix`
3. `make test`
4. then run django checks:
   ```
   make testshell
   cd webapp-django/
   ./manage.py check
   # skim the settings differences
   ./manage.py diffsettings --unified
   ```
5. `make run` then process some crashes and click around in the user interface
   1. verify home and product home work
   2. verify supersearch works
   3. verify report view works
   4. verify signature report works
   5. log in and verify the admin works

We can test more out on stage, but that's the bulk of things and if that works, we've tested database connectivity, caching, mozilla-django-oidc authentication, rate limiting, django-pipeline stuff, etc.
